styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -504,124 +507,51 @@ public void close() throws IOException {
-
-      // We found a pod? Bail.
-      if (pod != null) {
-        return;
-      }
-
-      // No pod found. Emit an error guarded by the state counter we are basing the error conclusion on.
-      stateManager.receiveIgnoreClosed(
-          Event.runError(workflowInstance, "No pod associated with this instance"), runState.counter());
-    });
-  }
-
-  private void pollPods() {
+  private void cleanupPods() {
     try {
-      try (Scope ss = tracer.spanBuilder("Styx.KubernetesDockerRunner.pollPods")
+      try (Scope ss = tracer.spanBuilder("Styx.KubernetesDockerRunner.cleanupPods")
           .setRecordEvents(true)
           .setSampler(Samplers.alwaysSample())
           .startScopedSpan()) {
-        tryPollPods();
+        tryCleanupPods();
       }
     } catch (Throwable t) {
-      LOG.warn("Error while polling pods", t);
+      LOG.warn("Error while cleaning pods", t);
     }
   }
 
+  /**
+   * Deletes stale workflow instance execution pods.
+   */
   @VisibleForTesting
-  synchronized void tryPollPods() {
-    // Fetch pods _before_ fetching all active states
-    final PodList list = client.pods().list();
-    final Map<WorkflowInstance, RunState> activeStates = stateManager.getActiveStates();
-      
-    // Note: Avoid blocking during the processing of the above fetched run states in order to prevent them
-    //       from going stale and losing races due to concurrent processing with other schedulers.
-      
-    examineRunningWFISandAssociatedPods(activeStates, list);
-
-    var podsToDelete = new ArrayList<String>();
-    for (Pod pod : list.getItems()) {
-      logEvent(Watcher.Action.MODIFIED, pod, list.getMetadata().getResourceVersion(), true);
-      final Optional<WorkflowInstance> workflowInstance = readPodWorkflowInstance(pod);
-      if (!workflowInstance.isPresent()) {
-        continue;
-      }
+  void tryCleanupPods() {
+    client.pods().list().getItems().stream()
+        .map(pod -> runAsync(guard(() -> tryCleanupPod(pod)), eventExecutor))
+        .collect(toList())
+        .forEach(CompletableFuture::join);
+  }
 
-      final RunState runState = activeStates.get(workflowInstance.get());
-      final boolean shouldDeletePod;
-      if (runState != null && isPodRunState(pod, runState)) {
-        emitPodEvents(pod, runState);
-        shouldDeletePod = shouldDeletePodWithRunState(workflowInstance.get(), pod, runState);
-      } else {
-        // The pod is stale as we fetched the active states _after_ listing all pods.
-        shouldDeletePod = shouldDeletePodWithoutRunState(workflowInstance.get(), pod);
-      }
-      if (shouldDeletePod) {
-        podsToDelete.add(pod.getMetadata().getName());
-      }
+  private void tryCleanupPod(Pod pod) {
+    var workflowInstance = readPodWorkflowInstance(pod);
+    if (workflowInstance.isEmpty()) {
+      return;
+    }
+    var runState = stateManager.getActiveState(workflowInstance.orElseThrow());
```

Comment:
Is `orElseThrow` a good practice?

Reply:
Yeah. Javadoc for `get`: `The preferred alternative to this method is {@link #orElseThrow()}.`